src/machi_file_proxy.erl

```diff
@@ -664,3 +659,26 @@ do_write(FHd, CsumTable, Filename, TaggedCsum, Offset, Size, Data) ->
             [Other, Filename, Offset, Size]),
             {error, Other}
     end.
+
+%% @doc Trim both right and left border of chunks to fit in to given
+%% range [LeftPos, RightPos]. TODO: write unit tests for this function.
+slice_both_side([], _, _) -> [];
+slice_both_side([{F, Offset, Chunk, _Csum}|L], LeftPos, RightPos)
+  when Offset < LeftPos andalso LeftPos < RightPos ->
+    TrashLen = 8 * (LeftPos - Offset),
+    <<_:TrashLen/binary, NewChunk/binary>> = Chunk,
+    NewChecksum = machi_util:make_tagged_csum(client_sha, Chunk),
+    NewH = {F, LeftPos, NewChunk, NewChecksum},
+    slice_both_side([NewH|L], LeftPos, RightPos);
+slice_both_side(Chunks, LeftPos, RightPos) when LeftPos =< RightPos ->
+    %% TODO: optimize
+    [{F, Offset, Chunk, _Csum}|L] = lists:reverse(Chunks),
+    Size = iolist_size(Chunk),
+    if RightPos < Offset + Size ->
+            NewSize = RightPos - Offset,
+            <<NewChunk:NewSize/binary, _/binary>> = Chunk,
+            NewChecksum = machi_util:make_tagged_csum(client_sha, Chunk),
```

Comment:
If the above ^^ sounds good, or if it needs some minor changes, then ... after fixing, do you have a suggestion for where we should make these definitions easy to find?  machi.hrl, or someplace else?  This GH ticket is probably not a good long-term place.